src/diffusers/models/unet_2d_blocks.py

```diff
@@ -1384,6 +1462,195 @@ def forward(
         return hidden_states, output_states
 
 
+class KDownBlock2D(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        temb_channels: int,
+        dropout: float = 0.0,
+        num_layers: int = 4,
+        resnet_eps: float = 1e-5, 
+        resnet_act_fn: str = "gelu",
+        resnet_groups: Optional[int] = None,
+        resnet_group_size: int = 32,
+        add_downsample=True,
+        resnet_time_scale_shift: str = "scale_shift",
+    ):
+        super().__init__()
+        resnets = []
+
+        for i in range(num_layers):
+            in_channels = in_channels if i == 0 else out_channels
+            resnets.append(
+                ResnetBlock2D(
+                    in_channels=in_channels,
+                    out_channels=out_channels,
+                    dropout=dropout,
+                    temb_channels=temb_channels,
+                    groups=resnet_groups,
+                    group_size=resnet_group_size,
+                    eps = resnet_eps,
+                    non_linearity = resnet_act_fn,
+                    norm1_scale_shift=True,
+                    time_embedding_norm=resnet_time_scale_shift,
+                    dropout_after_conv=True,
+                    conv_shortcut_bias=False,            
+                )
+            )
+
+        self.resnets = nn.ModuleList(resnets)
+
+        if add_downsample:
+             # YiYi's comments- might be able to use FirDownsample2D, look into details later
+            self.downsamplers = nn.ModuleList([KDownsample2d()])
+        else:
+            self.downsamplers = None
+
+        self.gradient_checkpointing = False
+
+    def forward(self, hidden_states, temb=None):
+        output_states = ()
+
+        if self.downsamplers is not None :
+            for downsampler in self.downsamplers:
+                hidden_states = downsampler(hidden_states)
+
+        for resnet in self.resnets:
+            if self.training and self.gradient_checkpointing:
+
+                def create_custom_forward(module):
+                    def custom_forward(*inputs):
+                        return module(*inputs)
+
+                    return custom_forward
+
+                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+            else:
+                hidden_states = resnet(hidden_states, temb)
+     
+        output_states += (hidden_states,) 
+
+        return hidden_states, output_states
+
+
+class KCrossAttnDownBlock2D(nn.Module):
```

Comment:
Do we need two types of blocks for the k-diffusion upsampler? Both K-Up and K-Down?

Reply:
I think one block definition is doable🤣 , we can get rid of the `attn1_types` and ` attn2_types` parameters too

in  k-diffusion, there is only 1 block definition, but it's used to create  3 types of blocks based on the `self_attn` and `cross_attn` parameters: (
(1) resnet only, 
(2) resnet + cross_attention, 
(3) resnet+ self_attention+ cross_attention
```python
class DBlock(layers.ConditionedSequential):
    def __init__(self, downsample=False, self_attn=False, cross_attn=False, ...):
        modules = [nn.Identity()]
        for i in range(n_layers):
            modules.append(ResConvBlock(...))
            if self_attn:
                modules.append(layers.SelfAttention2d(..))
            if cross_attn:
                modules.append(...))
        super().__init__(*modules)
        self.set_downsample(downsample)
```
https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/models/image_v1.py#L30

since we don't want to add these parameters to unet either, I think we can get around it by creating one block definition but three down/up block_types __names__ 

so for example, in our unet config, we can have 
```python
{
"up_block_types": [
 "K-Up",
  "K-Cross-Up",
  "K-Cross-Up",
   "K-Self-Cross-Up"
   ],
}
```
and in `get_down_block` we call the `K-Up` with different parameters based on the `up_block_type`
```python
get_down_block():
    if up_block_type == "K-Up":
        return K-Up(self_attn=False, cross_attn=False)
   elif up_block_type == "K-Cross-Up":
       return K-Up(self_attn=False, cross_attn=True)
   elif up_block_type == "K-Self-Cross-Up":
       return K-Up(self_attn=True, cross_attn=True)
```

Reply:
Actually I think it's fine as is :-) No need to merge the two classes